lacommunaute/openid_connect/middleware.py

```diff
@@ -10,14 +16,24 @@ def process_request(self, request):
             return
 
         query_params = request.GET.copy()
-        query_params.pop("proconnect_login")
-        new_url = (
-            f"{request.path}?{urlencode({k: v for k, v in query_params.items() if v})}"
-            if query_params
-            else request.path
-        )
+        [token] = query_params.pop("proconnect_login")
+
+        try:
+            decoded_data = jwt.decode(
+                token,
+                key=settings.PRO_CONNECT_AUTO_LOGIN_KEY,
+                algorithms=["HS256"],
+                # TODO: Remove once https://github.com/jpadilla/pyjwt/issues/939 is fixed
+                options={"verify_iat": False},
+            )
+            email = decoded_data["email"]
+        except jwt.InvalidTokenError:
+            logger.warning("Invalid proconnect_login token")
+            email = None
+
+        new_url = f"{request.path}?{query_params.urlencode()}" if query_params else request.path
 
-        if not request.user.is_authenticated:
-            return HttpResponseRedirect(reverse("openid_connect:authorize") + f"?next={new_url}")
+        if email is None or request.user.is_authenticated and request.user.email == email:
```

Comment:
Selon la solution retenue pour masquer l’email, il sera peut-être nécessaire d’utiliser une comparaison en temps constant pour éviter de permettre à un attaquant de deviner l’email dans un jwt.
```suggestion
        if email is None or request.user.is_authenticated and constant_time_compare(request.user.email, email):
```

Reply:
L’expiration très courte me semble être une limite suffisante, oui.

cc @Scttpr :

Pour Nexus, nous avons besoin de lier le même compte utilisateur (proconnect, identifié par l’email) entre plusieurs services (mettons communauté, emplois).
- Nous pourrions passer l’email de l’utilisateur dans un paramètre GET, mais il atterrirait dans nos logs. À éviter.
- On a choisi de stocker l’email dans un JWT chiffré, avec une expiration à 2 minutes, et de passer le JWT à l’autre service. L’autre service lit le contenu et fait un redirect vers proconnect avec `?login_hint=email`. Le risque s’en retrouve réduit à une interception des logs par un attaquant et un rejeu dans les deux minutes, qui exposerait l’email de l’utilisateur.

Est-ce acceptable ?